compiler/optimizer/abstractinterpreter/InliningMethodSummary.cpp

```diff
@@ -0,0 +1,179 @@
+      case Kind::NullCheckFolding:
+         return "NullCheck Folding";
+      case Kind::InstanceOfFolding:
+         return "InstanceOf Folding";
+      default:
+         TR_ASSERT_FATAL(false, "Unexpected Kind");
+      }
+   }
+
+bool TR::PotentialOptimizationVPPredicate::holdPartialOrderRelation(TR::VPConstraint* valueConstraint, TR::VPConstraint* testConstraint)
+   {
+   if (testConstraint->asIntConstraint()) // partial relation for int constraint
+      {
+      if (testConstraint->getLowInt() <= valueConstraint->getLowInt() && testConstraint->getHighInt() >= valueConstraint->getHighInt())
+         return true;
+      else
+         return false;
+      }
+   else if (testConstraint->asClassPresence()) // partial relation for nullness
+      {
+      if (testConstraint->isNonNullObject() && valueConstraint->isNonNullObject())
+         return true;
+      else if (testConstraint->isNullObject() && valueConstraint->isNullObject())
+         return true;
+      else
+         return false;
+      }
+   else if (testConstraint->asClassType()) // testing for checkcast
+      {
+      TR_ASSERT_FATAL(testConstraint->getClassType()->asResolvedClass(), "testConstraint unexpectedly admits unresolved class type");
+      if (valueConstraint->isNullObject())
+         return true; // VPClassType always accepts null
+
+      if (valueConstraint->isClassObject() == TR_yes)
+         return false; // valueConstraint->getClass() doesn't mean the right thing
+
+      TR_OpaqueClassBlock *valueClass = valueConstraint->getClass();
+      if (valueClass == NULL)
+         return false; // unknown type
+
+      TR_YesNoMaybe isSubtype = _vp->fe()->isInstanceOf(
+         valueClass, testConstraint->getClass(), valueConstraint->isFixedClass(), true);
+
+      return isSubtype == TR_yes;
+      }
+   else if (testConstraint->asClass()) // partial relation for instanceof
+      {
+      TR_ASSERT_FATAL(testConstraint->isClassObject() != TR_yes, "testConstraint unexpectedly admits class object");
+      TR_ASSERT_FATAL(testConstraint->getClass() != NULL, "testConstraint class unexpectedly admits null");
+      TR_ASSERT_FATAL(testConstraint->isNonNullObject(), "testConstraint unexpectedly admits null");
+      TR_ASSERT_FATAL(testConstraint->getPreexistence() == NULL, "testConstraint has unexpected pre-existence info");
+      TR_ASSERT_FATAL(testConstraint->getArrayInfo() == NULL, "testConstraint has unexpected array info");
+      TR_ASSERT_FATAL(testConstraint->getObjectLocation() == NULL, "testContraint has an unexpected location");
+      if (valueConstraint->isNullObject())
+         return true;
+
+      if (valueConstraint->isNonNullObject() && valueConstraint->isClassObject() != TR_yes && valueConstraint->getClass())
+         {
+         TR_YesNoMaybe yesNoMaybe = _vp->fe()->isInstanceOf(valueConstraint->getClass(), testConstraint->getClass(), valueConstraint->isFixedClass(), true);
+
+         if (yesNoMaybe != TR_maybe)
```

Comment:
Thanks for the explanation of what's going on with the positive and negative cases here. I don't think I followed the table, but this code definitely makes sense along with:
> So for the instanceOf folding test, we will apply optimization for both positive and negative predicate, but for the checkcast folding test, we will only apply optimization for the positive predicate.

While this isn't the separation of the positive and negative cases that we had discussed before with Karim and Dino, I think that at this point that's fine. The project can make those changes later as needed.

Since I've had to ask you for a few other unrelated changes anyway, would you mind just making this slightly more precise in the negative case? When `yesNoMaybe == TR_no`, we don't need to require that `valueConstraint->isNonNullObject()`, since at runtime the reference will be either null or a (non-null) instance of an incompatible type, and the result of `instanceof` is false either way. (When `yesNoMaybe == TR_yes` OTOH, we do still need to keep requiring that `valueConstraint->isNonNullObject()`.)

Reply:
Hi @jdmpapin , I have revised SPDX line for all newly added files and revised commit message.
The code for instanceOf folding in `InliningMethodSummary.cpp` has been revised to the same style as checkcast folding that you have shown in the last a few review comment. I have also specified the return for the `yesNoMaybe` condition as you required.

Please check it out. Thank you!